channel/channel.ml

```diff
@@ -50,66 +52,82 @@ module Make(Flow:V1_LWT.FLOW) = struct
   let ibuf_refill t =
     Flow.read t.flow >>= function
     | `Ok buf ->
+      (* users of get_ibuf (and therefore ibuf_refill) expect the buffer
+         returned here to have length >0; if Flow.read ever gives us empty
+         buffers, this will be violated causing Channel users to see Cstruct
+         exceptions *)
       t.ibuf <- Some buf;
       return_unit
-    | `Error _ | `Eof ->
-      fail Closed
+    | `Error e ->
+      fail (Read_error (Flow.error_message e))
+    | `Eof ->
+      (* close the flow before throwing exception; otherwise it will never be
+         GC'd *)
```

Comment:
This sounds like it might be another bug. Why doesn't the flow get GC'd eventually? Can we close forgotten flows automatically using `Gc.finalise` or something (as a last resort, logging a warning)?